prov/verbs/src/verbs_eq.c

```diff
@@ -144,30 +232,293 @@ fi_ibv_eq_cm_getinfo(struct fi_ibv_fabric *fab, struct rdma_cm_event *event,
 
 	connreq->handle.fclass = FI_CLASS_CONNREQ;
 	connreq->id = event->id;
+
+	if (fi_ibv_using_xrc()) {
+		ret = fi_ibv_eq_set_xrc_info(event, &connreq->xrc);
+		if (ret)
+			goto err3;
+	}
+
 	(*info)->handle = &connreq->handle;
 	fi_freeinfo(hints);
 	return 0;
+
+err3:
+	free(connreq);
 err2:
 	fi_freeinfo(*info);
 err1:
 	fi_freeinfo(hints);
 	return ret;
 }
 
+static inline int fi_ibv_eq_copy_event_data(struct fi_eq_cm_entry *entry,
+				size_t max_dest_len, const void *priv_data,
+				size_t priv_datalen)
+{
+	size_t datalen = MIN(max_dest_len - sizeof(*entry), priv_datalen);
+	if (datalen)
+		memcpy(entry->data, priv_data, datalen);
+
+	return datalen;
+}
+
+#ifdef INCLUDE_VERBS_XRC
+
+static void fi_ibv_eq_skip_xrc_cm_data(const void **priv_data,
+				       size_t *priv_data_len)
+{
+	const struct fi_ibv_xrc_cm_data *cm_data = *priv_data;
+
+	*priv_data = (cm_data + 1);
+	*priv_data_len -= sizeof(*cm_data);
+}
+
+static int
+fi_ibv_eq_xrc_connreq_event(struct fi_ibv_eq *eq, struct fi_eq_cm_entry *entry,
+			    const void **priv_data, size_t *priv_datalen)
+{
+	struct fi_ibv_connreq *connreq = container_of(entry->info->handle,
+						struct fi_ibv_connreq, handle);
+	struct fi_ibv_ep *ep;
+	struct fi_ibv_xrc_cm_data cm_data;
+	int ret;
+
+	if (!connreq->xrc.is_reciprocal) {
+		fi_ibv_eq_skip_xrc_cm_data(priv_data, priv_datalen);
+		return FI_SUCCESS;
+	}
+
+	/*
+	 * Reciprocal connections are initiated and handled internally by
+	 * the provider, get the endpoint that issued the original connection
+	 * request.
+	 */
+	ep = fi_ibv_eq_xrc_conn_tag2ep(eq, connreq->xrc.conn_tag);
+	if (!ep) {
+		VERBS_WARN(FI_LOG_FABRIC,
+			   "Reciprocal XRC connection tag not found\n");
+		goto send_reject;
+	}
+	ep->tgt_id = connreq->id;
+	ep->tgt_info = fi_dupinfo(entry->info);
+	ep->tgt_id->context = &ep->util_ep.ep_fid.fid;
+	ep->info->handle = entry->info->handle;
+
+	ret = rdma_migrate_id(ep->tgt_id, ep->eq->channel);
+	if (ret) {
+		VERBS_WARN(FI_LOG_FABRIC, "Could not migrate CM ID\n");
+		goto err;
+	}
+
+	ret = fi_ibv_accept_xrc(ep, FI_IBV_RECIP_CONN, &cm_data,
+				sizeof(cm_data));
+	if (ret) {
+		VERBS_WARN(FI_LOG_FABRIC,
+			   "Reciprocal XRC Accept failed %d\n", ret);
+		goto err;
+	}
+	return -FI_EAGAIN;
```

Comment:
This seems strange. I would have expected to return FI_SUCCESS. Is the FI_EAGAIN code necessary to move the state machine forward?

Reply:
This is the reciprocal connection to make the XRC data path bidirectional. It is created without interaction from and transparent to the application (i.e. no event is passed back to the application), so yes the FI_EAGAIN is just to keep the state machine moving forward.

Reply:
Should this be added as a comment?

Reply:
Yes, I'll add a comment. Thanks.

Reply:
Done. Thanks.